cmds/boot/boot.go

```diff
@@ -0,0 +1,403 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// boot allows to handover a system running linuxboot/u-root
+// to a legacy preinstalled operating system by replacing the traditional
+// bootloader path
+
+//
+// Synopsis:
+//	boot
+//
+// Description:
+//	If returns to u-root shell, the code didn't found a local bootable option
+//
+// Notes:
+//	The code is looking for boot/grub/grub.cfg file as to identify the
+//	boot option.
+//	The first bootable device found in the block device tree is the one used
+//	Windows is not supported (that is a work in progress)
+//
+// Example:
+//	boot -v 	- Start the script in verbose mode for debugging purpose
+
+package main
+
+import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
+	"github.com/u-root/u-root/pkg/kexec"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"flag"
+	"syscall"
+)
+
+const (
+	bootableMBR = 0xaa55
+	signatureOffset = 510
+)
+
+var verbose bool
+type options struct {
+        verbose bool
+}
+
+func check(e error) {
+	if e != nil {
+		log.Fatal(e)
+	}
+}
+
+func blkDevicesList(blkpath string, devpath string) []string {
+	var blkDevices []string
+	files, err := ioutil.ReadDir(blkpath)
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, file := range files {
+		deviceEntry, err := ioutil.ReadDir(blkpath + file.Name() + devpath)
+		if err != nil {
+			if ( verbose ) {
+				println("can t read directory")
+			}
+			continue
+		}
+		blkDevices = append(blkDevices, deviceEntry[0].Name())
+	}
+	return (blkDevices)
+}
+
+// checkForBootableMBR is looking for bootable MBR signature 
+// Current support is limited to Hard disk devices and USB devices
+func checkForBootableMBR(path string) (bool,error) {
+	var sig uint16
+	f, err := os.Open(path)
+	if err != nil {
+		return false, err
+	}
+	if err := binary.Read(io.NewSectionReader(f, signatureOffset, 2), binary.LittleEndian, &sig); err != nil {
+		return false,err
+	}
+	if sig != bootableMBR {
+		return false, nil
+	}
+	return true,nil	
+}
+
+// getDevicePartList returns all devices attached to a specific name like /dev/sdaX where X can move from 0 to 127
+// FIXME no support for devices which are included into subdirectory within /dev
+func getDevicePartList(path string) []string {
+	var returnValue []string
+	files, err := ioutil.ReadDir("/dev/")
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, file := range files {
+		if strings.HasPrefix(file.Name(), path) {
+			// We shall not return full device name
+			if file.Name() != path {
+				// We shall check that the remaining part is a number
+				returnValue = append(returnValue, file.Name())
+			}
+		}
+	}
+	return returnValue
+}
+
+// getSupportedFilesystem returns all block file system supported by the linuxboot kernel
```

Comment:
I think this is an easier way to do this

import (
	"fmt"
	"log"
	"strings"
	"io/ioutil"
)

func main() {
	fs, err := ioutil.ReadFile("/proc/filesystems") // note: open/close handling done by ReadFile
	if err != nil {
		log.Fatal(err)
	}

	for _, f := range strings.Split(string(fs), "\n") {
		n := strings.Fields(f)
		if len(n) != 1 {
			continue
		}
	fmt.Println(n[0])
	}
}

Reply:
Agreed. I have cut/paste your code